### ISSUE TITLE
Move to Finch

### ISSUE THREAD

**Issue by josevalim** [elixir-broadway/broadway_sqs#73]

**Comment by slashmili** [comment]
Hi,

What's needed to make the switch to Finch? This project is using [ex_aws_sqs](https://github.com/ex-aws/ex_aws_sqs/issues/36) which relies on [hackney v1.9.0](https://hex.pm/packages/hackney/1.9.0). The issue is that [ex_aws_sqs](https://hex.pm/packages/ex_aws_sqs) has not been updated for 3 years. 

Do you foresee that in this project we drop `ex_aws_sqs` dependency or the support for Finch needs to be added to `ex_aws_sqs` first? If the plan is to keep using `ex_aws_sqs`, We could add support for Finch to the upstream. However due to inactivity of maintainers, I'm not sure if our changes will be reviewed. 


In our business we are using Broadway and specifically `broadway_sqs` and willing to do the work to make this happen. Just need to ping right people to get the changes be merged in `ex_aws_sqs`

**Comment by josevalim** [comment]
We can either implement what we need from ex_aws_sqs here (I believe we already contributed a fast parser upstream in the past) or contribute it upstream. Both are fine.

There is also a separate library called `aws_credentials` that takes care of boring authentication bits. Another option is to ship `req_sqs`. cc @wojtekmach

**Comment by wojtekmach** [comment]
Right, I think ex_aws_sqs could get a Finch adapter so it works out of the box. Or, BroadwaySQS gets one instead.

That said, I think dropping ex_aws_sqs dependency seems like a good option too, we only use a couple of functions anyway and we could vendor in that tricky xml parser and depend on aws_credentials. On the other hand, unless we add a contract, people wouldn't be able to swap http clients.

I'm happy to look into req_sqs though on principle it'd be good to bring in as least deps as possible to broadway_sqs (transitive deps req would bring are jason (I plan to drop it in a couple minor versions), mime, and soon server_sent_events).

I think path of least resistance is adding :finch to :broadway, keeping :ex_aws_sqs dep, and implementing a BroadwaySQS.FinchClient.

**Comment by wojtekmach** [comment]
If we drop ex_aws_sqs we'd have to re-implement retries (unless we use req or similar)

**Comment by slashmili** [comment]
> I think path of least resistance is adding :finch to :broadway, keeping :ex_aws_sqs dep, and implementing a BroadwaySQS.FinchClient.

That is assuming we get to change `ex_aws_sqs` and make it's HTTP client adaptable to external adaptors?

The only challenge is getting our changes to be reviewed and merged into `ex_aws_sqs`. Otherwise it's a decent solution and requires minimal change in `ex_aws_sqs`!

**Comment by josevalim** [comment]
I can probably get us access for `ex_aws_sqs`. Perhaps transfer to this org but I think a port is worth considering in any case, especially in mind of issues like https://github.com/elixir-broadway/broadway_sqs/issues/88.

**Comment by slashmili** [comment]
Alright just to understand, the idea is to drop dependency to `ex_aws_sqs`(and by the extend  `ex_aws`)  in this repo?

This way:
* We implement the generic auth and what not that `ex_aws` does.
* We implement AWS API calls which `ex_aws_sqs` and `ex_aws` do.
* We use Finch(or req) for API calls.


BTW I looked at `ex_aws` and it already supports HTTP adapter in its config. It support hackney and req at this moment. I think it support JSON as well but I'm not sure. So it's possible to keep using `ex_aws` but with Req and JSON interface

**Comment by josevalim** [comment]
> BTW I looked at ex_aws and it already supports HTTP adapter in its config. It support hackney and req at this moment. I think it support JSON as well but I'm not sure. So it's possible to keep using ex_aws but with Req and JSON interface

That works for me!

The other option is to use `aws_credentials` to build a minimal SQS with Req. @wojtekmach do you have any example of req talking to AWS APIs somewhere?

**Comment by wojtekmach** [comment]
Here's a full example with Req + aws_credentials: https://gist.github.com/wojtekmach/ead58b0989e80c142ca90563d0e24d37

**Comment by halostatue** [comment]
Someone has done an update that at least removes the Hackney 1.x dependency  from `ex_aws_sqs`: https://github.com/BeamLabEU/beamlab_ex_aws_sqs

Switching to that (it uses `:beamlabs_ex_aws_sqs` as the application, but keeps the `ExAws.SQS` code namespace; I opened https://github.com/BeamLabEU/beamlab_ex_aws_sqs/issues/1 on this, because it would be trivial to use that).

---

(I've been working on a fork/PR that does this, but there's a bit more to it than appears at the beginning because of the forced switch to JSON.)

**Comment by mhooverdev** [comment]
Any updates on this issue? We are seeing conflicts with `stripity_stripe` needing `hackney` 4.0, and we cannot upgrade because we use `broadway` as well.

**Comment by slashmili** [comment]
> Someone has done an update that at least removes the Hackney 1.x dependency from ex_aws_sqs:

I don't think switching to `beamlabs_ex_aws_sqs` is a good solution as it we need to use yet another dependency that could be not updated in few months/years. 


> Any updates on this issue? We are seeing conflicts with stripity_stripe needing hackney 4.0, and we cannot upgrade because we use broadway as well.

IMO this issue is a different topic. It might solve the problem with deps versioning by switching to a different http client but we shouldn't push it because of that.

**Comment by josevalim** [comment]
Please try the main branch, req client has been merged in #90 thanks to @slashmili!
